Documentation/design-docs/dllmap.md

```diff
@@ -0,0 +1,246 @@
+
+### Callbacks
+
+As explained above, runtime will rise two dll specific events on each load attempt:
+* 1st callback - when loading a non-system library
+* 2nd callback - when finding an entrypoint
+
+Events will be defined in `AssemblyLoadContext` in `Private.CoreLib`. Default handlers that subscribe dll load events will implement the mono-based dllmap logic. They will take string as argument and return IntPtr of target libraries and entrypoints based on the parsed XML configuration file.
+
+Handlers implementation will stay in `AssemblyLoadContext`. Load library resolver will cache all the dll mapping results that got resolved (as key-value: IntPtr-hmod pairs). Thanks to that, the same library won't get loaded multiple times. User’s code will be able to subscribe to events and implement any loading behavior. That will give a user full flexibility when using dllmap and won’t limit defining the mapping to only xml-based style. Callbacks will be executed for non-system assemblies only. `System.Private.CoreLib` and other system assemblies will not receive callbacks, because that could cause problems like an infinite dll-mapping loop.
+
+### Resolution flow
+
+**User’s code [managed code]**
+
+-   Subscribes to `LoadLibraryModuleViaCallback` and `GetEntrypointViaCallback` events with their default or custom handler
+-   Uses DllImport directive and does the p/invoke
+	 ```c++
+	[DllImport("MyLibrary.dll", EntryPoint="MyFunction")]
+	static  extern  int  MyFunction();
+	…
+	System.Runtime.Loader.AssemblyLoadContext.Default.ResolveNativeDllName += LoadNativeDllViaDllMap;
+	System.Runtime.Loader.AssemblyLoadContext.Default.ResolveNativeEntrypointName += GetEntrypointViaDllMap;
+	…
+
+	MyFunction();
+	```
+
+**Runtime [unmanaged code]**
+-   Calls `LoadMappedLibrary` that raises `LoadLibraryModuleViaCallback` and `GetEntrypointViaCallback` events. Runtime uses a lock when it runs into a callback and releases the lock once the callback is done, to avoid infinite looping
+    
+**AssemblyLoadContext [unmanaged code]**
+-   Defines `LoadLibraryModuleViaCallbackand` and exposes an API:
+	```c++
+	IntPtr  LoadLibraryModuleViaCallback(string libraryName)
+	IntPtr  GetEntrypointViaCallback(string entrypointName, HMOD hmod)	
+	```
+
+**Dotnet.corefx.labs.dll [managed code]**
+-   Implements LoadNativeDllViaDllMap and GetEntrypointViaDllMap
+	```c++
+	IntPtr  LoadNativeDllViaDllMap(string libraryName)
+	{
+		if (libraryName in cachedResults)
+			return cachedResults[libraryName];
+		if (!mapStructure)
+			mapStructure = ReadAndParseXML();
+		targetLibraryName = mapStructure.GetLibrary(libraryName);
+		hmod = LoadLibrary(targetLibraryName);
+		AddToCache(libraryName, hmod);
+		return hmod;
+	}
+	```
+	```c++
+	IntPtr  GetEntrypointViaDllMap(string entrypointName, HMOD hmod)
+	{
+		targetEntrypointName = mapStructure.GetEntrypoint(entrypointName);
+		pvTarget = GetProcAddress(targetEntrypointName, hmod);
+		return pvTarget;
+	}
+	```
+
+## Testing
```

Comment:
Could you give a full example on how to use new dllmap? especially what  customer need to do to use dllmap??

Reply:
For sure it's necessary, but I would wait with this until we agree on the final shape of API  (to include exact names of events and handlers that we will provide).